### PR TITLE
LP vaults Deprecation and sdk v48 support

### PR DESCRIPTION
**Pull request by arhamchordia**
## 1. Overview

This PR contains changes for making LP vaults compatible with Osmosis sdk48 v24.0.0.
And deprecation related code for force unbind and force claim.

## 2. Implementation details

1. There were small changes in structure and their names in the modules.
2. Those changes in turn broke our decoding.
3. Changes in those struct name changes and ackBody struct makes it compatible again with Osmosis v21.
4. Addition of new function like `force_unbond` and `force_claim`

## 3. How to test/use

- There is demos/lp-vault-osmosis-upgrade directory : 
    - `sh setup.sh` to run the initial setup script
    - `sh contract_setup.sh` to deploy contracts and test them with a single bond
    - contract_setup.sh contains more commented out commands to test agains more accounts.
- There is demos/force_unbond_and_claim directory : 
    - `sh setup.sh` to setup Quasar, Osmosis and a relayer with contract deployed and a few bonds to start with.
    - `sh force_unbond_and_claim.sh` to test force unbind and claim.

## 4. Checklist

- [ ] Does the Readme need to be updated?

## 5. Limitations (optional)

*This breaks test tube tests as it does not have v24 released yet.*

### CODE REVIEW

File: smart-contracts/contracts/lp-strategy/src/ibc.rs [hunk 1]
```diff
@@ -297,13 +299,14 @@ pub fn handle_succesful_ack(
...
-    _ack_bin: Binary,
-    _pkt: &IbcPacketAckMsg,
+    ack_bin: Binary,
+    pkt: &IbcPacketAckMsg,
     mut pending: PendingBond,
     transferred_amount: Uint128,
 ) -> Result<Response, ContractError> {
     // once the ibc transfer to the ICA account has succeeded, we send the join pool message
     // we need to save and fetch
+    println!("ibc_pkt : {:?}", pkt);
```
Comment (akure) [inline smart-contracts/contracts/lp-strategy/src/ibc.rs:309]: println to be removed for production code.

File: smart-contracts/contracts/basic-vault/src/multitest/vault.rs [hunk 2]
```diff
@@ -17,7 +17,7 @@ fn try_bond() {
     assert_eq!(
         err,
         VaultContractError::Std(cosmwasm_std::StdError::GenericErr {
-            msg: "alloc::string::String not found".to_string()
+            msg: "type: alloc::string::String; key: [69, 63, 71, 5F, 63, 68, 61, 6E, 6E, 65, 6C] not found".to_string()
```
Comment (akure) [inline smart-contracts/contracts/basic-vault/src/multitest/vault.rs:20]: @arhamchordia What are these numbers?

File: smart-contracts/contracts/lp-strategy/src/ibc.rs [hunk 3]
```diff
@@ -363,11 +369,13 @@ pub fn handle_icq_ack(
 ) -> Result<Response, ContractError> {
     // todo: query flows should be separated by which flowType we're doing (bond, unbond, startunbond)
 
-    let ack: InterchainQueryPacketAck = from_binary(&ack_bin)?;
-    let resp: CosmosResponse = CosmosResponse::decode(ack.data.0.as_ref())?;
+    let buf: InterchainQueryPacketAckData = from_json(ack_bin)?;
+    let resp: CosmosResponse = CosmosResponse::decode(buf.data.0.as_ref())?;
+
+    println!("{:?}", resp);
```
Comment (akure) [inline smart-contracts/contracts/lp-strategy/src/ibc.rs:375]: println to be removed from production code.

**Review by akure — COMMENTED** [review]
@arhamchordia  Could you also summarise here in brief which struct were broken and needed to change?

File: demos/lp-vault-osmosis-upgrade/contract_setup.sh [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+CHAIN_ID=quasar
+
+INIT1='{"lock_period":6,"pool_id":1,"pool_denom":"gamm/pool/1","base_denom":"uosmo","local_denom":"ibc/ED07A3391A112B175915CD8FAF43A2DA8E4790EDE12566649D0C2F97716B8518","quote_denom":"stake","return_source_channel":"channel-0","transfer_channel":"channel-0","expected_connection":"connection-0"}'
+INIT2='{"lock_period":6,"pool_id":3,"pool_denom":"gamm/pool/3","base_denom":"uosmo","local_denom":"ibc/ED07A3391A112B175915CD8FAF43A2DA8E4790EDE12566649D0C2F97716B8518","quote_denom":"fakestake","return_source_channel":"channel-0","transfer_channel":"channel-0","expected_connection":"connection-0"}'
+INIT3='{"lock_period":6,"pool_id":2,"pool_denom":"gamm/pool/2","base_denom":"uosmo","local_denom":"ibc/ED07A3391A112B175915CD8FAF43A2DA8E4790EDE12566649D0C2F97716B8518","quote_denom":"usdc","return_source_channel":"channel-0","transfer_channel":"channel-0","expected_connection":"connection-0"}'
+
+cd ../../smart-contracts
+
+echo "Running store code"
+RES=$(quasarnoded tx wasm store artifacts/lp_strategy-aarch64.wasm --from alice --keyring-backend test -y --output json -b block --chain-id $CHAIN_ID --gas auto --fees 10000uqsr)
```
Comment (akure) [inline demos/lp-vault-osmosis-upgrade/contract_setup.sh:14]: Pls. use this in a variable lp_strategy-aarch64.wasm, for linux user :)

For linux users it should be set to lp_strategy.wasm and for mac users it should set to lp_strategy-aarch64.wasm

File: smart-contracts/contracts/lp-strategy/src/ibc.rs [hunk 5]
```diff
@@ -363,11 +369,13 @@ pub fn handle_icq_ack(
 ) -> Result<Response, ContractError> {
     // todo: query flows should be separated by which flowType we're doing (bond, unbond, startunbond)
 
-    let ack: InterchainQueryPacketAck = from_binary(&ack_bin)?;
-    let resp: CosmosResponse = CosmosResponse::decode(ack.data.0.as_ref())?;
+    let buf: InterchainQueryPacketAckData = from_json(ack_bin)?;
```
Comment (akure) [inline smart-contracts/contracts/lp-strategy/src/ibc.rs:372]: Havn't gone through the internals, but look like in a new  version ack_bin is coming as json object? 
We can then call this variable as ack_json.

Comment (arhamchordia) [reply smart-contracts/contracts/lp-strategy/src/ibc.rs:372]: Sure.

File: smart-contracts/contracts/basic-vault/src/multitest/vault.rs (hunk 2 above)
Comment (arhamchordia) [reply smart-contracts/contracts/basic-vault/src/multitest/vault.rs:20]: Keys are in bytes and it was trying to find this key in a map.

File: demos/lp-vault-osmosis-upgrade/contract_setup.sh (hunk 4 above)
Comment (arhamchordia) [reply demos/lp-vault-osmosis-upgrade/contract_setup.sh:14]: Sure. I will put an if else condition in it.

File: smart-contracts/contracts/basic-vault/src/multitest/vault.rs (hunk 2 above)
Comment (njerschow) [reply smart-contracts/contracts/basic-vault/src/multitest/vault.rs:20]: These bytes translate to icq_channel

**Review by LaurensKubat — COMMENTED** [review]
lgtm, but total understanding of the process is relatively low, a live demo/discussion would be really appreciated before we use this

**Pull request closed without merging**